### PR TITLE
Fix missing typing errors, improved strictness of typings

### PR DESCRIPTION
**Pull request by crutchcorn**
These changes are made to fix some of the TS inconsistencies raised in amwmedia/plop#143 as well as do more to provide both more accurate definition files as well as export useful interfaces and types that might be useful outside of `node-plop` (such as the downstream `plop` project for it's plopfiles).

I've noticed that 3 tests seem to be failing, however these issues seem to occur even on `master`, so I'm unsure they're related to the changes seen in this merge request (although I've been known to be wrong on this in the past 😅 Please feel free to double check this!). I'm more than willing to take a deeper look at those failing tests but want to make sure this isn't known and expected for the current developmental changes before moving forward on investigating that.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -1,4 +1,6 @@
 import inquirer = require("inquirer");
+import {IOptions as GlobOptions} from 'glob';
```
Comment (tannerntannern) [inline index.d.ts]: You're going to want the types from `'globby'`, not `'glob'`, as that's what plop uses.  Try this instead:

```typescript
// @types/globby doesn't export types for GlobOptions, so we have to work a little bit to extract them:
// GlobOptions is the second parameter of the sync function, which can be extracted with the Parameters<T> type
import {sync as _sync} from 'globby';
type GlobOptions = Parameters<typeof _sync>[1];
```

Comment (crutchcorn) [reply index.d.ts]: Oooh, okay. Gotcha, my bad. Admittedly haven't used the glob lib with TS projects much. Should have this updated by tonight

Comment (tannerntannern) [reply index.d.ts]: All good, man.  This comes from my old, abandoned attempt at what you're doing.  Glad to see someone else leading the charge!

**Comment by crutchcorn** [comment]
Updated PR with suggestions from @tannerntannern (tried giving credit as much as possible :) )

**Comment by amwmedia** [comment]
released to npm in `v0.17.4`

**Pull request merged**